biz.aQute.resolve/src/biz/aQute/resolve/ResolverValidator.java

```diff
@@ -65,21 +70,36 @@ public void setSystem(Resource resource) throws Exception {
 	}
 
 	public List<Resolution> validate() throws Exception {
+		FixedIndexedRepo repository = getRepository();
+		Set<Resource> resources = getAllResources(repository);
+		return validateResources(repository, resources);
+	}
+
+	public List<Resolution> validate(Collection<Resource> toBeChecked) throws Exception {
+		Set<Resource> allResources = new LinkedHashSet<Resource>();
+		for (URI uri : repositories) {
+			allResources.addAll(XMLResourceParser.getResources(uri));
+		}
+		allResources.addAll(toBeChecked);
+		ResourcesRepository repository = new ResourcesRepository(allResources);
+		return validateResources(repository, toBeChecked);
+	}
 
+	FixedIndexedRepo getRepository() throws MalformedURLException, URISyntaxException {
 		FixedIndexedRepo repository = new FixedIndexedRepo();
 		repository.setLocations(Strings.join(repositories));
-
-		Set<Resource> resources = getAllResources(repository);
-		setProperty("-runfw", "dummy");
-
-		return validateResources(repository, resources);
+		return repository;
 	}
 
-	public List<Resolution> validateResources(Repository repository, Set<Resource> resources) throws Exception {
+	public List<Resolution> validateResources(Repository repository, Collection<Resource> resources) throws Exception {
+		setProperty("-runfw", "dummy");
 		List<Resolution> result = new ArrayList<>();
-		for (Resource resource : resources) {
+		List<Resource> resourceList = new ArrayList<>(resources);
```

Comment:
Since you are regularly removing the first element of the list as well as random elements in the list with the removeAll, I suspect a LinkedList would be better than an ArrayList.